File: workers/decifra_workers/sources/price_live.py
"""Preço live (§3/§7) — SÓ on-demand (botão "atualizar"), nunca em batch (§10).

Connector económico sobre Google Shopping via DataForSEO (login+password) ou
Bright Data (token); SerpApi (SERPAPI_KEY) como alternativa. Escolhe a fonte mais
barata configurada. Devolve ofertas atuais como NormalizedOffer. Graceful: lista
vazia em falha/sem credenciais. Cada parser é puro e testável.
"""
from __future__ import annotations

import os

import httpx
from tenacity import retry, retry_if_exception_type, stop_after_attempt, wait_exponential

from ..config import Settings
from ..models import NormalizedOffer


class _Retryable(Exception):
    pass


def _currency(price_text: str | None) -> str | None:
    if not price_text:
        return None
    if "€" in price_text:
        return "EUR"
    if "£" in price_text:
        return "GBP"
    if "$" in price_text:
        return "USD"
    return None


def _num(v) -> float | None:
    if v is None:
        return None
    try:
        return float(str(v).replace(",", "."))
    except (TypeError, ValueError):
        return None


def parse_serpapi(payload: dict) -> list[NormalizedOffer]:
    offers: list[NormalizedOffer] = []
    for r in payload.get("shopping_results") or []:
        price = _num(r.get("extracted_price"))
        if price is None:
            continue
        offers.append(
            NormalizedOffer(
                store_name=r.get("source") or r.get("seller") or "Loja",
                price=price,
                currency=_currency(r.get("price")) or "EUR",
                url_affiliate=r.get("product_link") or r.get("link"),
                in_stock=True,
            )
        )
    return offers


def parse_dataforseo(payload: dict) -> list[NormalizedOffer]:
    """Mapeia uma resposta tipo DataForSEO Merchant/Google Shopping → ofertas.
    Best-effort; confirmar os nomes dos campos contra o schema vivo."""
    items = payload.get("items") or []
    offers: list[NormalizedOffer] = []
    for it in items:
        price = _num(it.get("price") or (it.get("price_info") or {}).get("current_price"))
        if price is None:
            continue
        offers.append(
            NormalizedOffer(
                store_name=it.get("seller") or it.get("shop") or it.get("source") or "Loja",
                price=price,
                currency=it.get("currency") or "EUR",
                url_affiliate=it.get("url") or it.get("link"),
                in_stock=it.get("availability", True) not in (False, "out_of_stock"),
            )
        )
    return offers


def parse_brightdata(payload: dict) -> list[NormalizedOffer]:
    rows = payload.get("results") or payload.get("data") or []
    offers: list[NormalizedOffer] = []
    for r in rows:
        price = _num(r.get("price") or r.get("final_price"))
        if price is None:
            continue
        offers.append(
            NormalizedOffer(
                store_name=r.get("seller") or r.get("merchant") or "Loja",
                price=price,
                currency=r.get("currency") or "EUR",
                url_affiliate=r.get("url") or r.get("link"),
                in_stock=True,
            )
        )
    return offers


class PriceLiveConnector:
    """Escolhe o provider mais barato configurado (DataForSEO > Bright Data > SerpApi)."""

    kind = "offers"

    def __init__(self, settings: Settings):
        self.settings = settings
        if settings.dataforseo_login and settings.dataforseo_password:
            self.provider = "dataforseo"
        elif settings.brightdata_token:
            self.provider = "brightdata"
        elif settings.serpapi_key:
            self.provider = "serpapi"
        else:
            self.provider = None

    @property
    def source_name(self) -> str:
        return self.provider or "price_live"

    def configured(self) -> bool:
        return self.provider is not None

    @retry(
        retry=retry_if_exception_type(_Retryable),
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, max=10),
        reraise=True,
    )
    def _get(self, *args, **kwargs) -> httpx.Response:
        resp = httpx.get(*args, timeout=30, **kwargs)
        if resp.status_code == 429 or resp.status_code >= 500:
            raise _Retryable(f"{resp.status_code}")
        resp.raise_for_status()
        return resp

    def fetch(self, ean: str | None, title: str | None) -> list[NormalizedOffer]:
        """Ofertas atuais para um EAN/título. Lista vazia em falha (graceful)."""
        if not self.configured():
            return []
        query = ean or title
        if not query:
            return []
        try:
            if self.provider == "serpapi":
                base = os.environ.get("SERPAPI_BASE_URL", "https://serpapi.com")
                resp = self._get(
                    f"{base}/search.json",
                    params={
                        "engine": "google_shopping",
                        "q": query,
                        "gl": "pt",
                        "hl": "pt",
                        "api_key": self.settings.serpapi_key,
                    },
                )
                return parse_serpapi(resp.json())
            if self.provider == "brightdata":
                base = os.environ.get("BRIGHTDATA_BASE_URL", "https://api.brightdata.com")
                resp = self._get(
                    f"{base}/serp/google/shopping",
                    params={"q": query, "country": "pt"},
                    headers={"Authorization": f"Bearer {self.settings.brightdata_token}"},
                )
                return parse_brightdata(resp.json())
            if self.provider == "dataforseo":
                base = os.environ.get("DATAFORSEO_BASE_URL", "https://api.dataforseo.com")
                resp = httpx.post(
                    f"{base}/v3/merchant/google/products/live/advanced",
                    auth=(self.settings.dataforseo_login or "", self.settings.dataforseo_password or ""),
                    json=[{"keyword": query, "location_code": 2620, "language_code": "pt"}],
                    timeout=30,
                )
                resp.raise_for_status()
                tasks = resp.json().get("tasks") or []
                result = (tasks[0].get("result") or [{}])[0] if tasks else {}
                return parse_dataforseo(result)
        except Exception:
            return []  # graceful: nunca quebra o pedido on-demand
        return []

File: web/src/lib/priceLive.ts
/**
 * Preço live on-demand (§3/§7) — caminho do botão "atualizar".
 *
 * Espelha o connector Python (workers/sources/price_live.py): Google Shopping via
 * DataForSEO / Bright Data / SerpApi (escolhe o mais barato configurado). SÓ
 * on-demand (nunca batch, §10). Graceful: sem credenciais devolve {configured:false};
 * em falha de rede devolve lista vazia (não quebra).
 */
import "server-only";
import { and, asc, eq } from "drizzle-orm";
import { getDb } from "@/db";
import { offers, productIdentifiers, products, sources, stores } from "@/db/schema";
import { invalidateProduct } from "./cache";

export type LiveOffer = {
  storeName: string;
  price: number;
  currency: string;
  url: string | null;
  inStock: boolean;
};

function currencyOf(priceText: string | undefined): string | null {
  if (!priceText) return null;
  if (priceText.includes("€")) return "EUR";
  if (priceText.includes("£")) return "GBP";
  if (priceText.includes("$")) return "USD";
  return null;
}

const numOf = (v: unknown): number | null => {
  if (v === null || v === undefined) return null;
  const n = typeof v === "number" ? v : Number(String(v).replace(",", "."));
  return Number.isFinite(n) ? n : null;
};

// ──────────────────────────────  parsers (puros)  ───────────────────────────
export function parseSerpapi(payload: { shopping_results?: unknown[] }): LiveOffer[] {
  const out: LiveOffer[] = [];
  for (const r of (payload.shopping_results ?? []) as Record<string, unknown>[]) {
    const price = numOf(r.extracted_price);
    if (price === null) continue;
    out.push({
      storeName: String(r.source ?? r.seller ?? "Loja"),
      price,
      currency: currencyOf(r.price as string) ?? "EUR",
      url: (r.product_link as string) ?? (r.link as string) ?? null,
      inStock: true,
    });
  }
  return out;
}

export function parseDataforseo(result: { items?: unknown[] }): LiveOffer[] {
  const out: LiveOffer[] = [];
  for (const it of (result.items ?? []) as Record<string, unknown>[]) {
    const price = numOf(it.price ?? (it.price_info as Record<string, unknown>)?.current_price);
    if (price === null) continue;
    out.push({
      storeName: String(it.seller ?? it.shop ?? it.source ?? "Loja"),
      price,
      currency: String(it.currency ?? "EUR"),
      url: (it.url as string) ?? (it.link as string) ?? null,
      inStock: it.availability !== false && it.availability !== "out_of_stock",
    });
  }
  return out;
}

export function parseBrightdata(payload: { results?: unknown[]; data?: unknown[] }): LiveOffer[] {
  const rows = (payload.results ?? payload.data ?? []) as Record<string, unknown>[];
  const out: LiveOffer[] = [];
  for (const r of rows) {
    const price = numOf(r.price ?? r.final_price);
    if (price === null) continue;
    out.push({
      storeName: String(r.seller ?? r.merchant ?? "Loja"),
      price,
      currency: String(r.currency ?? "EUR"),
      url: (r.url as string) ?? (r.link as string) ?? null,
      inStock: true,
    });
  }
  return out;
}

// ───────────────────────────────  provider  ─────────────────────────────────
type Provider = "dataforseo" | "brightdata" | "serpapi";

function selectProvider(): Provider | null {
  if (process.env.DATAFORSEO_LOGIN && process.env.DATAFORSEO_PASSWORD) return "dataforseo";
  if (process.env.BRIGHTDATA_TOKEN) return "brightdata";
  if (process.env.SERPAPI_KEY) return "serpapi";
  return null;
}

async function fetchLiveOffers(
  provider: Provider,
  ean: string | null,
  title: string | null,
): Promise<LiveOffer[]> {
  const query = ean || title;
  if (!query) return [];
  try {
    if (provider === "serpapi") {
      const base = process.env.SERPAPI_BASE_URL ?? "https://serpapi.com";
      const url = `${base}/search.json?engine=google_shopping&gl=pt&hl=pt&q=${encodeURIComponent(query)}&api_key=${process.env.SERPAPI_KEY}`;
      const res = await fetch(url);
      if (!res.ok) return [];
      return parseSerpapi(await res.json());
    }
    if (provider === "brightdata") {
      const base = process.env.BRIGHTDATA_BASE_URL ?? "https://api.brightdata.com";
      const res = await fetch(`${base}/serp/google/shopping?country=pt&q=${encodeURIComponent(query)}`, {
        headers: { Authorization: `Bearer ${process.env.BRIGHTDATA_TOKEN}` },
      });
      if (!res.ok) return [];
      return parseBrightdata(await res.json());
    }
    // dataforseo
    const base = process.env.DATAFORSEO_BASE_URL ?? "https://api.dataforseo.com";
    const auth = Buffer.from(`${process.env.DATAFORSEO_LOGIN}:${process.env.DATAFORSEO_PASSWORD}`).toString("base64");
    const res = await fetch(`${base}/v3/merchant/google/products/live/advanced`, {
      method: "POST",
      headers: { Authorization: `Basic ${auth}`, "content-type": "application/json" },
      body: JSON.stringify([{ keyword: query, location_code: 2620, language_code: "pt" }]),
    });
    if (!res.ok) return [];
    const data = await res.json();
    const result = data?.tasks?.[0]?.result?.[0] ?? {};
    return parseDataforseo(result);
  } catch {
    return [];
  }
}

// ───────────────────────────────  helpers DB  ───────────────────────────────
async function ensureSourceId(name: string): Promise<string> {
  const db = getDb();
  const [e] = await db.select({ id: sources.id }).from(sources).where(eq(sources.name, name)).limit(1);
  if (e) return e.id;
  const [c] = await db
    .insert(sources)
    .values({ name, kind: "offers", trustWeight: "0.6" })
    .returning({ id: sources.id });
  return c.id;
}

async function ensureStoreId(name: string): Promise<string> {
  const db = getDb();
  const [e] = await db.select({ id: stores.id }).from(stores).where(eq(stores.name, name)).limit(1);
  if (e) return e.id;
  const [c] = await db.insert(stores).values({ name }).returning({ id: stores.id });
  return c.id;
}

export type RefreshResult =
  | { configured: false }
  | { configured: true; found: false }
  | { configured: true; found: true; offers: { store: string | null; price: number | null; currency: string }[]; checkedAt: string };

/** Atualiza as offers de um produto via preço live (só a fonte live; mantém os feeds). */
export async function refreshProductPrice(slug: string): Promise<RefreshResult> {
  const provider = selectProvider();
  if (!provider) return { configured: false };

  const db = getDb();
  const [product] = await db
    .select({ id: products.id, name: products.canonicalName })
    .from(products)
    .where(eq(products.slug, slug))
    .limit(1);
  if (!product) return { configured: true, found: false };

  const [eanRow] = await db
    .select({ v: productIdentifiers.idValue })
    .from(productIdentifiers)
    .where(and(eq(productIdentifiers.productId, product.id), eq(productIdentifiers.idType, "ean")))
    .limit(1);

  const live = await fetchLiveOffers(provider, eanRow?.v ?? null, product.name);
  const sourceId = await ensureSourceId(provider);

  // Substitui apenas as offers desta fonte live (não toca nos feeds Awin).
  await db.delete(offers).where(and(eq(offers.productId, product.id), eq(offers.sourceId, sourceId)));
  for (const o of live) {
    const storeId = await ensureStoreId(o.storeName);
    await db.insert(offers).values({
      productId: product.id,
      storeId,
      price: String(o.price),
      currency: o.currency,
      urlAffiliate: o.url,
      inStock: o.inStock,
      sourceId,
      capturedAt: new Date(),
    });
  }

  await invalidateProduct(slug);

  const updated = await db
    .select({ store: stores.name, price: offers.price, currency: offers.currency })
    .from(offers)
    .leftJoin(stores, eq(stores.id, offers.storeId))
    .where(eq(offers.productId, product.id))
    .orderBy(asc(offers.price));

  return {
    configured: true,
    found: true,
    offers: updated.map((u) => ({ store: u.store, price: numOf(u.price), currency: u.currency })),
    checkedAt: new Date().toISOString(),
  };
}
